File: server/src/db/queryHelper.js
import { query } from './pool.js';

const SAFE_IDENTIFIER = /^[a-zA-Z_][a-zA-Z0-9_]*$/;

/**
 * Validate that a SQL identifier (table or column name) is safe.
 * Only allows letters, digits, and underscores — no special characters.
 */
const assertIdentifier = (name, label = 'identifier') => {
  if (!SAFE_IDENTIFIER.test(name)) {
    throw new Error(`Invalid ${label}: ${name}`);
  }
};

/**
 * SELECT rows from a table with optional WHERE conditions.
 * @param {string} table  - Table name
 * @param {object} conditions - Key/value pairs for WHERE clause (ANDed)
 * @param {string} orderBy - Column to ORDER BY (default 'id')
 * @returns {Promise<Array>} Matched rows
 */
export const findAll = async (table, conditions = {}, orderBy = 'id') => {
  assertIdentifier(table, 'table name');
  assertIdentifier(orderBy, 'orderBy column');

  const keys = Object.keys(conditions);
  keys.forEach((k) => assertIdentifier(k, 'column name'));
  const values = Object.values(conditions);

  let text = `SELECT * FROM ${table}`;
  if (keys.length > 0) {
    const whereClauses = keys.map((key, i) => `${key} = $${i + 1}`);
    text += ` WHERE ${whereClauses.join(' AND ')}`;
  }
  text += ` ORDER BY ${orderBy}`;

  return query(text, values);
};

/**
 * SELECT a single row by its primary key.
 * @param {string} table - Table name
 * @param {number|string} id - Primary key value
 * @returns {Promise<object|undefined>} The row, or undefined
 */
export const findById = async (table, id) => {
  assertIdentifier(table, 'table name');
  const text = `SELECT * FROM ${table} WHERE id = $1`;
  const rows = await query(text, [id]);
  return rows[0];
};

/**
 * INSERT a row and return the created record.
 * @param {string} table - Table name
 * @param {object} dataObject - Column/value pairs to insert
 * @returns {Promise<object>} The inserted row
 */
export const insertRow = async (table, dataObject) => {
  assertIdentifier(table, 'table name');
  const keys = Object.keys(dataObject);
  keys.forEach((k) => assertIdentifier(k, 'column name'));
  const values = Object.values(dataObject);
  const placeholders = keys.map((_, i) => `$${i + 1}`);

  const text = `INSERT INTO ${table} (${keys.join(', ')}) VALUES (${placeholders.join(', ')}) RETURNING *`;
  const rows = await query(text, values);
  return rows[0];
};

/**
 * UPDATE a row by id and return the updated record.
 * @param {string} table - Table name
 * @param {number|string} id - Primary key value
 * @param {object} dataObject - Column/value pairs to update
 * @returns {Promise<object|undefined>} The updated row, or undefined
 */
export const updateRow = async (table, id, dataObject) => {
  assertIdentifier(table, 'table name');
  const keys = Object.keys(dataObject);
  keys.forEach((k) => assertIdentifier(k, 'column name'));
  const values = Object.values(dataObject);

  const setClauses = keys.map((key, i) => `${key} = $${i + 1}`);
  const text = `UPDATE ${table} SET ${setClauses.join(', ')} WHERE id = $${keys.length + 1} RETURNING *`;
  const rows = await query(text, [...values, id]);
  return rows[0];
};

/**
 * DELETE a row by id and return the deleted record.
 * @param {string} table - Table name
 * @param {number|string} id - Primary key value
 * @returns {Promise<object|undefined>} The deleted row, or undefined
 */
export const deleteRow = async (table, id) => {
  assertIdentifier(table, 'table name');
  const text = `DELETE FROM ${table} WHERE id = $1 RETURNING *`;
  const rows = await query(text, [id]);
  return rows[0];
};

File: backend-dotnet/src/Incentive.Infrastructure/Data/QueryHelper.cs
using System.Data;
using System.Text.RegularExpressions;
using Dapper;

namespace Incentive.Infrastructure.Data;

/// <summary>
/// Generic CRUD helper using Dapper, matching the Node.js db/queryHelper.js contract.
/// All table/column names are validated against a safe identifier pattern.
/// </summary>
public partial class QueryHelper
{
    private readonly DbConnectionFactory _db;

    [GeneratedRegex(@"^[a-zA-Z_][a-zA-Z0-9_]*$")]
    private static partial Regex SafeIdentifierRegex();

    public QueryHelper(DbConnectionFactory db) => _db = db;

    private static void AssertIdentifier(string name, string label = "identifier")
    {
        if (!SafeIdentifierRegex().IsMatch(name))
            throw new ArgumentException($"Invalid {label}: {name}");
    }

    /// <summary>
    /// SELECT * FROM table WHERE conditions ORDER BY orderBy.
    /// </summary>
    public async Task<IEnumerable<dynamic>> FindAllAsync(
        string table,
        Dictionary<string, object>? conditions = null,
        string orderBy = "id")
    {
        AssertIdentifier(table, "table name");
        AssertIdentifier(orderBy, "orderBy column");

        var sql = $"SELECT * FROM {table}";
        var parameters = new DynamicParameters();

        if (conditions is { Count: > 0 })
        {
            var clauses = new List<string>();
            var i = 0;
            foreach (var (key, value) in conditions)
            {
                AssertIdentifier(key, "column name");
                clauses.Add($"{key} = @p{i}");
                parameters.Add($"p{i}", value);
                i++;
            }
            sql += $" WHERE {string.Join(" AND ", clauses)}";
        }

        sql += $" ORDER BY {orderBy}";

        using var conn = await _db.CreateConnectionAsync();
        return await conn.QueryAsync(sql, parameters);
    }

    /// <summary>
    /// SELECT * FROM table WHERE id = @id.
    /// </summary>
    public async Task<dynamic?> FindByIdAsync(string table, int id)
    {
        AssertIdentifier(table, "table name");
        using var conn = await _db.CreateConnectionAsync();
        return await conn.QueryFirstOrDefaultAsync($"SELECT * FROM {table} WHERE id = @id", new { id });
    }

    /// <summary>
    /// INSERT INTO table (...) VALUES (...) RETURNING *.
    /// </summary>
    public async Task<dynamic?> InsertRowAsync(string table, Dictionary<string, object?> data)
    {
        AssertIdentifier(table, "table name");
        var keys = data.Keys.ToList();
        keys.ForEach(k => AssertIdentifier(k, "column name"));

        var columns = string.Join(", ", keys);
        var paramNames = string.Join(", ", keys.Select(k => $"@{k}"));

        var sql = $"INSERT INTO {table} ({columns}) VALUES ({paramNames}) RETURNING *";

        var parameters = new DynamicParameters();
        foreach (var (key, value) in data)
            parameters.Add(key, value);

        using var conn = await _db.CreateConnectionAsync();
        return await conn.QueryFirstOrDefaultAsync(sql, parameters);
    }

    /// <summary>
    /// UPDATE table SET ... WHERE id = @id RETURNING *.
    /// </summary>
    public async Task<dynamic?> UpdateRowAsync(string table, int id, Dictionary<string, object?> data)
    {
        AssertIdentifier(table, "table name");
        var keys = data.Keys.ToList();
        keys.ForEach(k => AssertIdentifier(k, "column name"));

        var setClauses = keys.Select((k, i) => $"{k} = @{k}");
        var sql = $"UPDATE {table} SET {string.Join(", ", setClauses)} WHERE id = @_id RETURNING *";

        var parameters = new DynamicParameters();
        foreach (var (key, value) in data)
            parameters.Add(key, value);
        parameters.Add("_id", id);

        using var conn = await _db.CreateConnectionAsync();
        return await conn.QueryFirstOrDefaultAsync(sql, parameters);
    }

    /// <summary>
    /// DELETE FROM table WHERE id = @id RETURNING *.
    /// </summary>
    public async Task<dynamic?> DeleteRowAsync(string table, int id)
    {
        AssertIdentifier(table, "table name");
        using var conn = await _db.CreateConnectionAsync();
        return await conn.QueryFirstOrDefaultAsync($"DELETE FROM {table} WHERE id = @id RETURNING *", new { id });
    }

    /// <summary>
    /// Execute a raw parameterized query and return typed results.
    /// </summary>
    public async Task<IEnumerable<T>> QueryAsync<T>(string sql, object? param = null)
    {
        using var conn = await _db.CreateConnectionAsync();
        return await conn.QueryAsync<T>(sql, param);
    }

    /// <summary>
    /// Execute a raw parameterized query and return dynamic results.
    /// </summary>
    public async Task<IEnumerable<dynamic>> QueryAsync(string sql, object? param = null)
    {
        using var conn = await _db.CreateConnectionAsync();
        return await conn.QueryAsync(sql, param);
    }

    /// <summary>
    /// Execute a raw parameterized query and return the first result or default.
    /// </summary>
    public async Task<T?> QueryFirstOrDefaultAsync<T>(string sql, object? param = null)
    {
        using var conn = await _db.CreateConnectionAsync();
        return await conn.QueryFirstOrDefaultAsync<T>(sql, param);
    }

    /// <summary>
    /// Execute a non-query SQL command. Returns affected row count.
    /// </summary>
    public async Task<int> ExecuteAsync(string sql, object? param = null)
    {
        using var conn = await _db.CreateConnectionAsync();
        return await conn.ExecuteAsync(sql, param);
    }
}
